src/commands/Validation.ts

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+type ValidationMatchExpression<Ok, Err> = { ok?: (ok: Ok) => any, err?: (err: Err) => any};
+
+/**
+ * Why do we need a Result Monad for the parser signiture.
+ * I (Gnuxie) don't like monadic error handling, simply because
+ * I'm a strong believer in failing early, yes i may be misinformed.
+ * The only reason we don't use an exception in this case is because
+ * these are NOT to be used nilly willy and thrown out of context
+ * from an unrelated place. The Monad ensures locality (in terms of call chain)
+ * to the user interface by being infuriating to deal with.
+ * It also does look different to an exception
+ * to a naive programmer. Ideally though, if the world had adopted
+ * condition based error handling i would simply create a condition
+ * type for validation errors that can be translated/overidden by
+ * the command handler, and it wouldn't have to look like this.
+ * It's important to remember the errors we are reporting are to do with user input,
+ * we're trying to tell the user they did something wrong and what that is.
```

Comment:
Well, generally speaking, there are several kinds of errors and error-handling:

1. Something in my code is broken. In JS, that's generally represented by  a `TypeError` (even if it's not a type error, sigh) or `SyntaxError` (except `SyntaxError` is also used when parsing JSON, sigh).
2. Something in my environment is broken. In JS, that's generally other cases of `Error`.
3. Something in my input is broken. There is nothing built-in in JS, but that could be handled by creating a new variant of `Error`.

While I'm ok with using monads in the right setting and there *are* ways to use Monads in JS with some syntactic support, by encoding them as a generator and its caller. However, introducing a different paradigm in the middle of an application feels a bit cumbersome, so I'd like to avoid it if possible.

Reply:
On the other hand, I don't see anything that looks like monadic binding, so this isn't really a monad, more a data structure to represent good parse result/bad parse result, right? That feels ok to me.

What does it bring wrt a new variant of `Error`, though? Better type-checking?

Reply:
> so this isn't really a monad, more a data structure to represent good parse result/bad parse result, right? That feels ok to me.

Yep

>What does it bring wrt a new variant of Error, though? Better type-checking?

Because as I said, so it doesn't get misused. The validation errors are a description of something that the user has done wrong, for the user to see.